source/particle/output/ascii.cc

```diff
@@ -0,0 +1,140 @@
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/output/ascii.h>
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Output
+    {
+      template <int dim>
+      ASCIIOutput<dim>::ASCIIOutput()
+        :
+        file_index(0)
+      {}
+
+      template <int dim>
+      std::string
+      ASCIIOutput<dim>::output_particle_data(const std::multimap<types::LevelInd, Particle<dim> > &particles,
+                                             const std::vector<std::pair<std::string, unsigned int> > &property_component_list,
+                                             const double /*time*/)
+      {
+        const std::string output_file_prefix = "particle-" + Utilities::int_to_string (file_index, 5);
+        const std::string output_path_prefix = this->get_output_directory() + output_file_prefix;
+        const std::string full_filename = output_path_prefix + "." + Utilities::int_to_string(Utilities::MPI::this_mpi_process(this->get_mpi_communicator()), 4) + ".txt";
+        std::ofstream output (full_filename.c_str());
+        if (!output)
+          std::cout << "ERROR: proc " << Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) << " could not create " << full_filename << std::endl;
+
+        // Print the header line
+        output << "# ";
+        for (unsigned int i = 0; i < dim; ++i)
+          output << "position[" << i << "] ";
+
+        output << "id ";
+
+        std::vector<std::pair<std::string,unsigned int> >::const_iterator property = property_component_list.begin();
+        for (; property!=property_component_list.end(); ++property)
+          {
+            // If it's a 1D element, print just the name, otherwise use []
+            if (property->second == 1)
+              {
+                output << property->first << " ";
+              }
+            else
+              {
+                for (unsigned int i=0; i<property->second; ++i) output << property->first << "[" << i << "] ";
+              }
+          }
+        output << "\n";
+
+        // And print the data for each particle
+        for (typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator it=particles.begin(); it!=particles.end(); ++it)
+          {
+            const Point<dim> position = it->second.get_location();
+
+            for (unsigned int i = 0; i < dim; ++i)
+              output << position[i] << " ";
+
+            output << it->second.get_id() << " ";
```

Comment:
here and elsewhere: you can micro-optimize by replacing
```
  output << " "
```
by
```
  output << ' '
```
This outputs the same single space character, but it avoids searching for the end of the string enclosed in double quotes and copying a variable number of characters.